core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -904,4 +916,82 @@ private static ListMultimap<PlanNodeId, ExchangeSourceHandle> getInputsForRemote
         }
         return result.build();
     }
+
+    private static class LoadedSplits
+    {
+        private final PlanNodeId planNodeId;
+        private final List<Split> splits;
+
+        private LoadedSplits(PlanNodeId planNodeId, List<Split> splits)
+        {
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.splits = ImmutableList.copyOf(requireNonNull(splits, "splits is null"));
+        }
+
+        public PlanNodeId getPlanNodeId()
+        {
+            return planNodeId;
+        }
+
+        public List<Split> getSplits()
+        {
+            return splits;
+        }
+    }
+
+    private static class SplitLoadingFuture
+            extends AbstractFuture<LoadedSplits>
+    {
+        private final PlanNodeId planNodeId;
+        private final SplitSource splitSource;
+        private final int splitBatchSize;
+        private final LongConsumer getSplitTimeRecorder;
+        @GuardedBy("this")
+        private final List<Split> loadedSplits = new ArrayList<>();
+        @GuardedBy("this")
+        private ListenableFuture<SplitBatch> currentSplitBatch = immediateFuture(null);
+
+        SplitLoadingFuture(PlanNodeId planNodeId, SplitSource splitSource, int splitBatchSize, LongConsumer getSplitTimeRecorder)
+        {
+            this.planNodeId = requireNonNull(planNodeId, "planNodeId is null");
+            this.splitSource = requireNonNull(splitSource, "splitSource is null");
+            this.splitBatchSize = splitBatchSize;
+            this.getSplitTimeRecorder = requireNonNull(getSplitTimeRecorder, "getSplitTimeRecorder is null");
+        }
+
+        // Called to initiate loading and to load next batch if not finished
+        public synchronized void load()
+        {
+            if (currentSplitBatch == null || splitSource.isFinished()) {
+                return;
+            }
+            checkState(currentSplitBatch.isDone(), "next batch of splits requested before previous batch is done");
+            currentSplitBatch = splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize);
+
+            long start = System.nanoTime();
+            addSuccessCallback(currentSplitBatch, splitBatch -> {
+                getSplitTimeRecorder.accept(start);
+                synchronized (this) {
+                    loadedSplits.addAll(splitBatch.getSplits());
+
+                    if (splitBatch.isLastBatch()) {
+                        splitSource.close();
+                        set(new LoadedSplits(planNodeId, loadedSplits));
+                    }
+                    else {
+                        load();
+                    }
+                }
+            });
```

Comment:
Better invoke a callback in an executor to avoid an unlimited recursion stack if the split sources returns already completed futures (for example when splits are loaded synchronously). I guess you can reuse the `@ForQueryExecution` executor scheduler already available in the context.

Reply:
I've made changes to use the `@ForScheduler` executor here as that was available in FaultTolerantDistributedStagesScheduler context